policy/engine.go

```diff
@@ -0,0 +1,784 @@
+	}
+	if err := e.validateNames(dnsNames, ips, emails, uris, []string{}); err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+// AreSANSAllowed verifies that all names in the slice of SANs are allowed.
+// The SANs are first split into DNS names, IPs, email addresses and URIs.
+func (e *NamePolicyEngine) AreSANsAllowed(sans []string) (bool, error) {
+	dnsNames, ips, emails, uris := x509util.SplitSANs(sans)
+	if err := e.validateNames(dnsNames, ips, emails, uris, []string{}); err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+// IsDNSAllowed verifies a single DNS domain is allowed.
+func (e *NamePolicyEngine) IsDNSAllowed(dns string) (bool, error) {
+	if err := e.validateNames([]string{dns}, []net.IP{}, []string{}, []*url.URL{}, []string{}); err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+// IsIPAllowed verifies a single IP domain is allowed.
+func (e *NamePolicyEngine) IsIPAllowed(ip net.IP) (bool, error) {
+	if err := e.validateNames([]string{}, []net.IP{ip}, []string{}, []*url.URL{}, []string{}); err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+// ArePrincipalsAllowed verifies that all principals in an SSH certificate are allowed.
+func (e *NamePolicyEngine) ArePrincipalsAllowed(cert *ssh.Certificate) (bool, error) {
+	dnsNames, ips, emails, usernames := splitPrincipals(cert.ValidPrincipals)
+	if err := e.validateNames(dnsNames, ips, emails, []*url.URL{}, usernames); err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+// appendSubjectCommonName appends the Subject Common Name to the appropriate slice of names. The logic is
+// similar as x509util.SplitSANs: if the subject can be parsed as an IP, it's added to the ips. If it can
+// be parsed as an URL, it is added to the URIs. If it contains an @, it is added to emails. When it's none
+// of these, it's added to the DNS names.
+func appendSubjectCommonName(subject pkix.Name, dnsNames *[]string, ips *[]net.IP, emails *[]string, uris *[]*url.URL) {
+	commonName := subject.CommonName
+	if commonName == "" {
+		return
+	}
+	if ip := net.ParseIP(commonName); ip != nil {
+		*ips = append(*ips, ip)
+	} else if u, err := url.Parse(commonName); err == nil && u.Scheme != "" {
+		*uris = append(*uris, u)
+	} else if strings.Contains(commonName, "@") {
+		*emails = append(*emails, commonName)
+	} else {
+		*dnsNames = append(*dnsNames, commonName)
+	}
+}
+
+// splitPrincipals splits SSH certificate principals into DNS names, emails and usernames.
+func splitPrincipals(principals []string) (dnsNames []string, ips []net.IP, emails, usernames []string) {
+	dnsNames = []string{}
+	ips = []net.IP{}
+	emails = []string{}
+	usernames = []string{}
+	for _, principal := range principals {
+		if strings.Contains(principal, "@") {
+			emails = append(emails, principal)
+		} else if ip := net.ParseIP(principal); ip != nil {
+			ips = append(ips, ip)
+		} else if len(strings.Split(principal, ".")) > 1 {
+			dnsNames = append(dnsNames, principal)
```

Comment:
Fixed this in 066bf320866de37f509ee647466b5c9a6666c8ff. 

I've added the IP range similar to x509 options now.

However, I think for Host certs, `EmailAddresses` and `Principals` shouldn't be checked. I think it makes more sense for users to configure DNS/IP permitted/denied constraints instead of them including those in the `Principals`. A principal `localhost` in `ValidPrincipals` in a Host cert is currently considered a DNS name in the policy engine; not a "plain" `Principal`. 

I think it can be made to work to check all three if required, but that would require a slightly different flow for `validateNames` as compared to the x509 one.